Applications/Spire/Source/Ui/FocusObserver.cpp

```diff
@@ -0,0 +1,129 @@
+#include "Spire/Ui/FocusObserver.hpp"
+#include <QApplication>
+#include <QFocusEvent>
+
+using namespace boost::signals2;
+using namespace Spire;
+
+namespace {
+  QObject* find_focus_event_filter_from_parent(QWidget& widget) {
+    auto parent = &widget;
+    while (parent) {
+      if(parent->isWindow()) {
+        break;
+      }
+      auto filter = parent->findChild<QObject*>("FocusEventFilter",
+        Qt::FindDirectChildrenOnly);
+      if(filter) {
+        return filter;
+      }
+      parent = parent->parentWidget();
+    }
+    return nullptr;
+  }
+}
+
+struct FocusObserver::FocusEventFilter : QObject {
+  FocusObserver* m_observer;
+  Qt::FocusReason m_focus_reason;
+
+  FocusEventFilter(FocusObserver& observer)
+      : QObject(const_cast<QWidget*>(observer.m_widget)),
```

Comment:
Remove the ```const_cast```.

Reply:
`observer.m_widget` is a `const` variable, but the copy constructor of `QObject` doesn't accept a `const` object.

Reply:
The event filter doesn't need a parent, and it can outlive the widget, so it can be removed.

Reply:
```QObject``` does not have a copy constructor. The fact that ```QObject```'s constructor is rejecting this should be an indication that the entire operation is incorrect.